File: src/types/node.ts
/**
 * Core node types for the context graph.
 * Every concept in the system is a typed node with identity and status.
 */

export type NodeType =
  | "goal"
  | "decision"
  | "constraint"
  | "task"
  | "risk"
  | "question"
  | "context"
  | "plan"
  | "note";

export type NodeStatus = "accepted" | "proposed" | "rejected" | "superseded";

export interface NodeId {
  /** Stable identifier for this node, survives rebases and merges */
  id: string;
  /** Optional namespace for organizing nodes */
  namespace?: string;
}

export interface TextRange {
  /** Start position in the source text */
  start: number;
  /** End position in the source text */
  end: number;
  /** Optional reference to the source file */
  source?: string;
}

/**
 * Types of relationships between nodes in the graph.
 */
export type RelationshipType =
  | "parent-child" // Hierarchical relationships (sub-decisions, subtasks)
  | "depends-on" // Dependencies (task dependencies, decision dependencies)
  | "references" // References (decisions referencing goals, tasks referencing decisions)
  | "supersedes" // Replacement relationships (new decision supersedes old)
  | "related-to" // General relationships
  | "implements" // Implementation relationships (task implements decision)
  | "blocks" // Blocking relationships (risk blocks task)
  | "mitigates"; // Mitigation relationships (task mitigates risk)

/**
 * A typed relationship between two nodes.
 */
export interface NodeRelationship {
  /** Type of relationship */
  type: RelationshipType;
  /** Target node ID */
  target: NodeId;
  /** Optional: reverse relationship type (for bidirectional relationships) */
  reverseType?: RelationshipType;
  /** Optional: metadata about this relationship */
  metadata?: {
    /** When this relationship was created */
    createdAt?: string;
    /** Who created this relationship */
    createdBy?: string;
    /** Optional description or context */
    description?: string;
  };
}

export interface NodeMetadata {
  /** When this node was created */
  createdAt: string;
  /** Who created this node */
  createdBy: string;
  /** When this node was last modified */
  modifiedAt: string;
  /** Who last modified this node */
  modifiedBy: string;
  /** Tags for categorization */
  tags?: string[];
  /** Git commit hash when this node was first accepted/implemented */
  implementedInCommit?: string;
  /** Git commit hashes that reference or update this node */
  referencedInCommits?: string[];
  /** Version number for optimistic locking (increments on each change) */
  version: number;
}

/**
 * Base interface for all context nodes.
 */
export interface ContextNode {
  /** Unique identifier */
  id: NodeId;
  /** Type of node */
  type: NodeType;
  /** Current status */
  status: NodeStatus;
  /**
   * Normalized short label for UI/display.
   * Optional for backward compatibility.
   */
  title?: string;
  /**
   * Normalized long-form description/body for UI/display.
   * Optional for backward compatibility.
   */
  description?: string;
  /** The semantic content of this node */
  content: string;
  /** Text range in source Markdown (if applicable) */
  textRange?: TextRange;
  /** Metadata */
  metadata: NodeMetadata;
  /** Typed relationships to other nodes (graph edges) */
  relationships?: NodeRelationship[];
  /** Related node IDs (deprecated: use relationships instead, kept for backward compatibility) */
  relations?: NodeId[];
  /** Nodes that reference this node (reverse relations, computed from relationships) */
  referencedBy?: NodeId[];
  /** Markdown files that contain this node */
  sourceFiles?: string[];
}

/**
 * A goal represents an objective or desired outcome.
 */
export interface GoalNode extends ContextNode {
  type: "goal";
  /** Optional success criteria */
  criteria?: string[];
}

/**
 * A decision represents a made choice with rationale.
 */
export interface DecisionNode extends ContextNode {
  type: "decision";
  /** The decision that was made */
  decision: string;
  /** Why this decision was made */
  rationale: string;
  /** Alternatives that were considered */
  alternatives?: string[];
  /** When this decision was made */
  decidedAt?: string;
}

/**
 * A constraint represents a limitation or requirement.
 */
export interface ConstraintNode extends ContextNode {
  type: "constraint";
  /** The constraint itself */
  constraint: string;
  /** Why this constraint exists */
  reason?: string;
}

/**
 * A task represents work to be done.
 */
export interface TaskNode extends ContextNode {
  type: "task";
  /** Current state of the task */
  state: "open" | "in-progress" | "blocked" | "completed" | "cancelled";
  /** Optional assignee */
  assignee?: string;
  /** Optional due date */
  dueDate?: string;
  /** Dependencies (other task IDs) */
  dependencies?: NodeId[];
}

/**
 * A risk represents a potential problem or concern.
 */
export interface RiskNode extends ContextNode {
  type: "risk";
  /** Severity: low, medium, high, critical */
  severity: "low" | "medium" | "high" | "critical";
  /** Likelihood: unlikely, possible, likely, certain */
  likelihood: "unlikely" | "possible" | "likely" | "certain";
  /** Mitigation strategy */
  mitigation?: string;
}

/**
 * A question represents an open question that needs an answer.
 */
export interface QuestionNode extends ContextNode {
  type: "question";
  /** The question being asked */
  question: string;
  /** Optional answer (if resolved) */
  answer?: string;
  /** When answered (if resolved) */
  answeredAt?: string;
}

/**
 * A plan represents a structured approach or roadmap.
 */
export interface PlanNode extends ContextNode {
  type: "plan";
  /** Steps or phases in the plan */
  steps?: PlanStep[];
}

export interface PlanStep {
  /** Description of the step */
  description: string;
  /** Related task or goal IDs */
  references?: NodeId[];
  /** Order/sequence */
  order: number;
}

/**
 * Union type of all node types.
 */
export type AnyNode =
  | GoalNode
  | DecisionNode
  | ConstraintNode
  | TaskNode
  | RiskNode
  | QuestionNode
  | ContextNode
  | PlanNode;

File: server/src/types/node.rs
//! Core node types for the context graph.
//! Mirrors src/types/node.ts.

use serde::{Deserialize, Serialize};

use crate::sensitivity::Sensitivity;

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum NodeType {
    Goal,
    Decision,
    Constraint,
    Task,
    Risk,
    Question,
    Context,
    Plan,
    Note,
}

impl NodeType {
    pub fn as_str(&self) -> &'static str {
        match self {
            NodeType::Goal => "goal",
            NodeType::Decision => "decision",
            NodeType::Constraint => "constraint",
            NodeType::Task => "task",
            NodeType::Risk => "risk",
            NodeType::Question => "question",
            NodeType::Context => "context",
            NodeType::Plan => "plan",
            NodeType::Note => "note",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum NodeStatus {
    Accepted,
    Proposed,
    Rejected,
    Superseded,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct NodeId {
    pub id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub namespace: Option<String>,
}

impl NodeId {
    pub fn key(&self) -> String {
        self.namespace
            .as_ref()
            .map(|n| format!("{}:{}", n, self.id))
            .unwrap_or_else(|| self.id.clone())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn node_id_key_without_namespace() {
        let id = NodeId {
            id: "a".to_string(),
            namespace: None,
        };
        assert_eq!(id.key(), "a");
    }

    #[test]
    fn node_id_key_with_namespace() {
        let id = NodeId {
            id: "goal-001".to_string(),
            namespace: Some("ui".to_string()),
        };
        assert_eq!(id.key(), "ui:goal-001");
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TextRange {
    pub start: u32,
    pub end: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum RelationshipType {
    ParentChild,
    DependsOn,
    References,
    Supersedes,
    RelatedTo,
    Implements,
    Blocks,
    Mitigates,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NodeRelationship {
    #[serde(rename = "type")]
    pub relationship_type: RelationshipType,
    pub target: NodeId,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reverse_type: Option<RelationshipType>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<RelationshipMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RelationshipMetadata {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub created_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub created_by: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NodeMetadata {
    pub created_at: String,
    pub created_by: String,
    pub modified_at: String,
    pub modified_by: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub implemented_in_commit: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub referenced_in_commits: Option<Vec<String>>,
    pub version: u32,
    /// Sensitivity classification for the node (default: internal).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sensitivity: Option<Sensitivity>,
    /// SHA-256 hash of node content at apply time (fingerprinting for IP protection).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content_hash: Option<String>,
    /// Source attribution for provenance tracking.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_attribution: Option<String>,
    /// IP classification for governance.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ip_classification: Option<String>,
    /// License identifier for content.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub license: Option<String>,
}

/// Context node: unified struct for all node types.
/// Type-specific fields (e.g. decision, rationale for Decision) are optional.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContextNode {
    pub id: NodeId,
    #[serde(rename = "type")]
    pub node_type: NodeType,
    pub status: NodeStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text_range: Option<TextRange>,
    pub metadata: NodeMetadata,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relationships: Option<Vec<NodeRelationship>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relations: Option<Vec<NodeId>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub referenced_by: Option<Vec<NodeId>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_files: Option<Vec<String>>,
    // Type-specific (decision)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub decision: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rationale: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub alternatives: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub decided_at: Option<String>,
    // Type-specific (task)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state: Option<TaskState>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub assignee: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub due_date: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dependencies: Option<Vec<NodeId>>,
    // Type-specific (risk)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub severity: Option<RiskSeverity>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub likelihood: Option<RiskLikelihood>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mitigation: Option<String>,
    // Type-specific (question)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub question: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub answer: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub answered_at: Option<String>,
    // Type-specific (constraint)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub constraint: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum TaskState {
    Open,
    InProgress,
    Blocked,
    Completed,
    Cancelled,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum RiskSeverity {
    Low,
    Medium,
    High,
    Critical,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum RiskLikelihood {
    Unlikely,
    Possible,
    Likely,
    Certain,
}
